{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8b2a5b6-5cc7-46bd-be48-2fb09166dcc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data manipulation\n",
    "# --------------------------------------------------------\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import math\n",
    "import time\n",
    "\n",
    "# Visualizations\n",
    "# --------------------------------------------------------\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import missingno as msno\n",
    "import altair as alt\n",
    "from plotnine import ggplot, geom_point, aes, stat_smooth, facet_wrap\n",
    "from bokeh.plotting import figure, show\n",
    "import seaborn.objects as so\n",
    "import plotly.express as px\n",
    "\n",
    "# Custom scripts\n",
    "# ---------------------------------------------------------\n",
    "from scripts.data_processing import cleaning, compare_and_drop_duplicates\n",
    "\n",
    "# Other\n",
    "# ---------------------------------------------------------\n",
    "from collections import Counter\n",
    "import umap\n",
    "from sklearn.preprocessing import MinMaxScaler, RobustScaler, QuantileTransformer, OrdinalEncoder\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bff607bf-b49e-46cd-b940-4a607e958ade",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load raw data\n",
    "var_info = pd.read_csv(\"./data/raw/variation_information.tsv\", sep=\"\\t\")\n",
    "cancer_clinvar = pd.read_csv(\"./data/raw/cancermama_clinvarmain.csv\", sep=\"\\t\", low_memory=False)\n",
    "upgenevsrep = pd.read_csv(\"./data/raw/UP.geneVsrepList.txt\", sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb7004a7-be38-4880-bec6-bf134d12c509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean each dataset\n",
    "var_info = cleaning(var_info, df_name=\"variation_information\")\n",
    "cancer_clinvar = cleaning(cancer_clinvar, df_name=\"cancermama_clinvarmain\")\n",
    "upgenevsrep = cleaning(upgenevsrep, df_name=\"UP.geneVsrepList\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4585509c-5e45-450f-8671-029f22d94c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join var_info and cancer_clinvar\n",
    "var_info[\"real_id\"] = var_info[\"rsID\"].astype(str) + \" \" + var_info[\"ClinInfo\"].astype(str)\n",
    "cancer_clinvar[\"real_id\"] = cancer_clinvar[\"snpId\"].astype(str) + \" \" + cancer_clinvar[\"ClinInfo\"].astype(str)\n",
    "cancer_clinvar = cancer_clinvar.drop([\"ClinInfo\"], axis=1)\n",
    "merged_df = cancer_clinvar.merge(var_info, left_on=\"real_id\", right_on=\"real_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f1b9b73-f51c-4c8a-9071-e0ce48293b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns 'clinSign' and 'ClinClass' have 2 differences.\n",
      "Columns '_clinSignCode' and 'Classification' have 2 differences.\n"
     ]
    }
   ],
   "source": [
    "# Remove duplicated columns with the same values\n",
    "merged_df = compare_and_drop_duplicates(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e51b2515-525f-46dc-8f3d-b44af3b6ab2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Binary classification using PG and LP as 1 \n",
    "merged_df['bin_class'] = merged_df['Classification'].apply(lambda x: 1 if x in ['pg'] else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81b7cff2-9948-4ebc-89ef-bc8b0a1c610b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop geneId as it has the same information of gene\n",
    "merged_df.drop([\"geneId\"], axis=1, inplace=True)\n",
    "# ids\n",
    "id_columns = [\"numSubmit\", \"_variantId\", \"real_id\", \"origName\", \"rcvAcc\", \"snpId\"]\n",
    "merged_df.drop(id_columns, axis=1, inplace=True)\n",
    "# url in column, no additional info\n",
    "url_columns = [\"phenotype\"]\n",
    "merged_df.drop(url_columns, axis=1, inplace=True)\n",
    "# _mouseOver have repeated information included in other columns:\n",
    "other_columns = [\"_mouseOver\"]\n",
    "merged_df.drop(other_columns, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bebb8eae-c2da-4563-9e03-265ed9fd4ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill nans per column\n",
    "for i in merged_df.columns:\n",
    "    # 1 missing value in clinSign and ClinClass for the same row\n",
    "    if i in [\"clinSign\", \"ClinClass\", \"molConseq\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"unkwown\")\n",
    "    # 45 missing values in phenotypeList\n",
    "    if i in [\"phenotypeList\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"not provided\")\n",
    "    # 45 missing values in rcvAcc\n",
    "    if i in [\"rcvAcc\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"rcv000000000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ac6a09d-775b-457b-a13d-3495aa62ff13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=========================\n",
      "\n",
      "#chrom\n",
      "chr17    49125\n",
      "chr2     30159\n",
      "chr16    29047\n",
      "chr11    24704\n",
      "chr5     24219\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "chromStart\n",
      "37025628    32\n",
      "47414419    32\n",
      "47414420    22\n",
      "95099770    17\n",
      "78131728    16\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "chromEnd\n",
      "37025630    33\n",
      "47414421    28\n",
      "37025608    19\n",
      "95099771    17\n",
      "78131619    16\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "name\n",
      "c>t    49778\n",
      "g>a    47477\n",
      "a>g    40605\n",
      "t>c    30231\n",
      "del    26620\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "score\n",
      "1    222786\n",
      "2     97300\n",
      "3     10212\n",
      "0      8501\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "reserved\n",
      "0,0,128        146341\n",
      "0,210,0        113196\n",
      "210,0,0         49066\n",
      "137,121,212     25602\n",
      "128,128,128      4594\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "blockSizes\n",
      "1    306825\n",
      "2     20478\n",
      "3      2890\n",
      "4      2190\n",
      "5      1184\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "clinSign\n",
      "uncertain significance                          146311\n",
      "likely benign                                    98087\n",
      "pathogenic                                       36150\n",
      "conflicting classifications of pathogenicity     25599\n",
      "benign                                            9485\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "reviewStatus\n",
      "criteria_provided_no_conflict    294495\n",
      "criteria_provided_conflict        25591\n",
      "reviewed                          10212\n",
      "not_classified                     8501\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "type\n",
      "single nucleotide variant    293049\n",
      "deletion                      24727\n",
      "duplication                   10786\n",
      "microsatellite                 3966\n",
      "indel                          3509\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "molConseq\n",
      "missense variant                 153726\n",
      "intron variant                    58942\n",
      "synonymous variant                45246\n",
      "frameshift variant                27494\n",
      "non-coding transcript variant     18329\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "testedInGtr\n",
      "n    338775\n",
      "y        24\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "phenotypeList\n",
      "hereditary cancer-predisposing syndrome    38084\n",
      "not provided                               21836\n",
      "fanconi anemia                             10093\n",
      "not specified                               8202\n",
      "neurofibromatosis,type 1                    7123\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "origin\n",
      "germ       325119\n",
      "unk         10246\n",
      "germsom      1217\n",
      "som           742\n",
      "germnov       405\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "cytogenetic\n",
      "13q13.1     18418\n",
      "11q22.3     16299\n",
      "5q22.2      14170\n",
      "17q21.31    13936\n",
      "17q11.2     13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "vcfDesc\n",
      "chr17:-1:na>na        49\n",
      "chr13:-1:na>na        26\n",
      "chr2:-1:na>na          9\n",
      "chr17:78131665:t>c     8\n",
      "chr17:78132279:g>a     8\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "_clinSignCode\n",
      "vus    146312\n",
      "lb     103710\n",
      "pg      36155\n",
      "cf      25602\n",
      "lp      12911\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "_originCode\n",
      "germ       326504\n",
      "unk         10246\n",
      "germsom      1307\n",
      "som           742\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "_allTypeCode\n",
      "subst     293057\n",
      "loss       24727\n",
      "gain       10786\n",
      "ins         6084\n",
      "seqlen      3966\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "simplified_hgvs\n",
      "c>t, g>a    38780\n",
      "a>g, t>c    23382\n",
      "c>t         19784\n",
      "g>a         19595\n",
      "a>g         18645\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "Gene\n",
      "brca2    18418\n",
      "atm      16299\n",
      "apc      14170\n",
      "brca1    13936\n",
      "nf1      13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "Start\n",
      "32315507     18418\n",
      "108223066    16299\n",
      "112707497    14170\n",
      "43044294     13936\n",
      "31094926     13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "End\n",
      "32400268     18418\n",
      "108369102    16299\n",
      "112846239    14170\n",
      "43170327     13936\n",
      "31377677     13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "ClinInfo\n",
      "nm_152468.5(tmc8):c.84g>a (p.arg28=)       8\n",
      "nm_152468.5(tmc8):c.187t>c (p.trp63arg)    8\n",
      "nm_152468.5(tmc8):c.185g>a (p.arg62his)    8\n",
      "nm_152468.5(tmc8):c.175c>t (p.gln59ter)    8\n",
      "nm_152468.5(tmc8):c.168g>a (p.ala56=)      8\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "ClinClass\n",
      "uncertain significance                          146311\n",
      "likely benign                                    98087\n",
      "pathogenic                                       36150\n",
      "conflicting classifications of pathogenicity     25599\n",
      "benign                                            9485\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "Classification\n",
      "vus    146312\n",
      "lb     103710\n",
      "pg      36155\n",
      "cf      25602\n",
      "lp      12911\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "bin_class\n",
      "0    302644\n",
      "1     36155\n",
      "Name: count, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in merged_df.columns:\n",
    "    print(\"=\"*25)\n",
    "    print()\n",
    "    print(merged_df[i].value_counts().nlargest(5))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "11f749d1-7920-4a2d-b078-9913db9638dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv(\"./data/interim/merged_df.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4cc888-b917-489f-8ce6-d723ee75f17a",
   "metadata": {},
   "source": [
    "Cluster -1 Genes: GLMN, GPC3, HFE, HRAS\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    GLMN: GLMN (Glomulin) gene mutations are associated with glomuvenous malformations.\n",
    "    GPC3: GPC3 (Glypican 3) is involved in cell growth control and differentiation and mutations can cause Simpson-Golabi-Behmel syndrome.\n",
    "    HFE: HFE gene mutations are linked to hereditary hemochromatosis, affecting iron metabolism.\n",
    "    HRAS: HRAS is an oncogene associated with the regulation of cell division and mutations can lead to cancer.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Genes related to various syndromes and cancers.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 0 Genes: RTEL1, RUNX1, SBDS, SDHA, SDHAF2, SDHB, SDHC, SDHD, SMAD4, SMARCA4, SMARCB1, SMC1A, SMC3, STK11, SUFU, SYNE1, SYNE2, TERT, TGFBR1, TGFBR2, TMEM127\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    RTEL1: Involved in DNA repair and telomere maintenance.\n",
    "    RUNX1: Key role in hematopoiesis and is associated with leukemia.\n",
    "    SBDS: Linked to Shwachman-Diamond syndrome.\n",
    "    SDHA, SDHAF2, SDHB, SDHC, SDHD: Part of the succinate dehydrogenase complex, mutations can cause mitochondrial disorders.\n",
    "    SMAD4, SMARCA4, SMARCB1: Tumor suppressor genes involved in cancer development.\n",
    "    STK11, SUFU: Tumor suppressor genes, mutations associated with Peutz-Jeghers syndrome and medulloblastoma respectively.\n",
    "    TERT, TGFBR1, TGFBR2: Associated with telomere function and TGF-beta signaling pathway.\n",
    "    SYNE1, SYNE2: Involved in nuclear envelope structure and mutations can cause muscular dystrophies.\n",
    "    TMEM127: Associated with pheochromocytoma.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Predominantly tumor suppressor genes and genes involved in syndromes or cancer.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 1 Genes: PIK3CA, PMS2, POLD1, POLH, POU6F2, PTCH1, PTEN, RAD51C, RAD51D, RB1, RET, RHBDF2, RHOA, RINT1, RNF43, ROS1, RTEL1, RUNX1, SBDS, SDHA, SDHAF2, SDHB, SDHC, SDHD, SMAD4, SMARCA4, SMARCB1, SMC1A, SMC3, STK11, SUFU, SYNE1, SYNE2, TERT, TGFBR1, TGFBR2, TMEM127\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    PIK3CA: Involved in the PI3K/AKT pathway, mutations associated with various cancers.\n",
    "    PMS2, POLD1, POLH: DNA repair genes, associated with Lynch syndrome.\n",
    "    POU6F2, PTCH1, PTEN: Involved in developmental processes and tumor suppression.\n",
    "    RAD51C, RAD51D: Involved in homologous recombination repair, mutations can lead to cancer.\n",
    "    RB1, RET, RHBDF2: Tumor suppressor genes, mutations associated with retinoblastoma and other cancers.\n",
    "    RHOA: Involved in signal transduction and cell migration.\n",
    "    RINT1, RNF43, ROS1: Associated with DNA repair and cancer pathways.\n",
    "    RTEL1, RUNX1, SBDS, SDHA, SDHAF2, SDHB, SDHC, SDHD: As above, involved in DNA repair, hematopoiesis, mitochondrial function.\n",
    "    SMAD4, SMARCA4, SMARCB1, SMC1A, SMC3, STK11, SUFU: Tumor suppressor genes.\n",
    "    SYNE1, SYNE2, TERT, TGFBR1, TGFBR2, TMEM127: As above, involved in nuclear structure, telomere function, and TGF-beta signaling.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Predominantly DNA repair genes and tumor suppressors, associated with various cancers and syndromes.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 2 Genes: CYP21A2, DDB2, DICER1, DIS3L2, EHBP1, EPCAM, ERCC1, ERCC2, ERCC3, ERCC4, ERCC5, FAN1, FANCA, FANCB, FANCC, FANCD2, FANCE, FANCF, FANCG, FANCI, FANCL, FANCM, FH, FLCN, FOS, FSTL3, FTH1, GATA2, GBE1, GCK\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    CYP21A2: Associated with congenital adrenal hyperplasia.\n",
    "    DDB2, DICER1, DIS3L2: Involved in DNA repair and RNA processing.\n",
    "    EHBP1, EPCAM: Associated with epithelial cell adhesion and migration.\n",
    "    ERCC1, ERCC2, ERCC3, ERCC4, ERCC5: Involved in nucleotide excision repair.\n",
    "    FAN1, FANCA, FANCB, FANCC, FANCD2, FANCE, FANCF, FANCG, FANCI, FANCL, FANCM: Involved in Fanconi anemia pathway.\n",
    "    FH, FLCN: Associated with metabolic disorders and cancer.\n",
    "    FOS, FSTL3: Involved in signal transduction and cell differentiation.\n",
    "    FTH1, GATA2, GBE1, GCK: Involved in iron storage, hematopoiesis, and glycogen metabolism.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Genes involved in DNA repair, metabolic disorders, and signal transduction.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 3 Genes: AIP, AKT1, ALK, ANTXR2, APC, ASCC1, ATM, ATR, BARD1, BCL2, BCL2L11, BLM, BMPR1A, BRCA1, BRCA2, BRIP1, BTK, CBL, CD79A, CD79B, CDH1, CDK4, CDK6, CDKN1B, CDKN2A, CEBPA, CHEK1, CHEK2, CIC, CLP1, CUL3, CYLD\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    AIP: Associated with pituitary adenomas.\n",
    "    AKT1, ALK: Involved in cell survival and proliferation.\n",
    "    ANTXR2: Associated with anthrax toxin receptor.\n",
    "    APC: Tumor suppressor gene, mutations associated with colorectal cancer.\n",
    "    ASCC1, ATM, ATR: Involved in DNA repair and signal transduction.\n",
    "    BARD1, BCL2, BCL2L11: Involved in apoptosis regulation.\n",
    "    BLM, BMPR1A: Involved in DNA repair and bone morphogenesis.\n",
    "    BRCA1, BRCA2, BRIP1: DNA repair genes, mutations associated with breast cancer.\n",
    "    BTK, CBL, CD79A, CD79B: Involved in B-cell receptor signaling.\n",
    "    CDH1, CDK4, CDK6, CDKN1B, CDKN2A: Tumor suppressor genes.\n",
    "    CEBPA, CHEK1, CHEK2, CIC, CLP1, CUL3, CYLD: Involved in cell cycle regulation and DNA repair.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Genes involved in DNA repair, apoptosis regulation, and cell cycle control.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 4 Genes: LYST, MC1R, MEN1, MET, MLH1, MLH3, MSH2, MSH6, MUTYH, NBN, NDUFA13, NF1, NF2, NFE2L2, NFKBIA, NFKBIE, NIN, NKX2-1, NLRP1, NOTCH1, NQO1, NRAS, NSD1, NTHL1, NTRK1, NTRK2, NTRK3\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "LYST: Associated with Chediak-Higashi syndrome.\n",
    "MC1R: Involved in pigmentation and associated with melanoma risk.\n",
    "MEN1, MET: Tumor suppressor genes.\n",
    "MLH1, MLH3, MSH2, MSH6: DNA mismatch repair genes.\n",
    "MUTYH, NBN, NDUFA13: Involved in DNA repair and mitochondrial function.\n",
    "NF1, NF2: Tumor suppressor genes.\n",
    "NFE2L2, NFKBIA, NFKBIE: Involved in oxidative stress response and inflammation.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "NIN, NKX2-1, NLRP1, ​NOTCH1: Involved in cell signaling and immune response.\n",
    "NQO1, NRAS, NSD1, NTHL1: Involved in cellular metabolism and DNA repair.\n",
    "NTRK1, NTRK2, NTRK3: Neurotrophic receptors, involved in nervous system development.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Genes involved in DNA repair, tumor suppression, and signaling pathways.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 5 Genes: FANCB, FANCC, FANCD2, FANCE, FANCF, FANCG, FANCI, FANCL, FANCM\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "FANCB, FANCC, FANCD2, FANCE, FANCF, FANCG, FANCI, FANCL, FANCM: All genes are part of the Fanconi anemia pathway, involved in DNA repair.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: All genes are involved in the Fanconi anemia pathway.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 6 Genes: ALK, ANTXR1, ANTXR2, APC, ASCC1, ATM, ATR, BARD1, BCL2, BCL2L11, BLM, BMPR1A, BRCA1, BRCA2, BRIP1, BTK, CBL, CD79A, CD79B, CDH1, CDK4, CDK6, CDKN1B, CDKN2A, CEBPA, CHEK1, CHEK2, CIC, CLP1, CUL3, CYLD\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Same as Cluster 3.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 7 Genes: KIF1B, KIT, KLHDC8B, LIG4, LYST, MAX, MC1R, MEN1, MET, MLH1, MLH3, MSH2, MSH6, MUTYH, NBN, NDUFA13, NF1, NF2, NFE2L2, NFKBIA, NFKBIE, NIN, NKX2-1, NLRP1, NOTCH1, NQO1, NRAS, NSD1, NTHL1, NTRK1, NTRK2, NTRK3\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Same as Cluster 4.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Cluster 8 Genes: MSH2, MSH6, MSMB, MSR1, MUTYH, NBN, NDUFA13, NF1, NF2, NFE2L2, NFKBIA, NFKBIE, NIN, NKX2-1, NLRP1, NOTCH1, NQO1, NRAS, NSD1, NTHL1, NTRK1, NTRK2, NTRK3\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Similarity: Same as Cluster 4.\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "172dd49f-8414-4351-b421-093f3592dad9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>labels</th>\n",
       "      <th>gene</th>\n",
       "      <th>counts</th>\n",
       "      <th>agg_kind</th>\n",
       "      <th>perc</th>\n",
       "      <th>similarity_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>GLMN</td>\n",
       "      <td>1</td>\n",
       "      <td>SINE/Alu</td>\n",
       "      <td>0.24</td>\n",
       "      <td>Genes related to various syndromes and cancers.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>2</td>\n",
       "      <td>LINE/L1, SINE/Alu</td>\n",
       "      <td>0.47</td>\n",
       "      <td>Genes related to various syndromes and cancers.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-1</td>\n",
       "      <td>HFE</td>\n",
       "      <td>2</td>\n",
       "      <td>LINE/L1, SINE/Alu</td>\n",
       "      <td>0.47</td>\n",
       "      <td>Genes related to various syndromes and cancers.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-1</td>\n",
       "      <td>HRAS</td>\n",
       "      <td>2</td>\n",
       "      <td>SINE/Alu, Simple_repeat</td>\n",
       "      <td>0.47</td>\n",
       "      <td>Genes related to various syndromes and cancers.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>RTEL1</td>\n",
       "      <td>1</td>\n",
       "      <td>LINE/L2</td>\n",
       "      <td>0.24</td>\n",
       "      <td>Predominantly tumor suppressor genes and genes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>10</td>\n",
       "      <td>NF1</td>\n",
       "      <td>2</td>\n",
       "      <td>SINE/Alu, SINE/MIR</td>\n",
       "      <td>0.47</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>10</td>\n",
       "      <td>NF2</td>\n",
       "      <td>3</td>\n",
       "      <td>LINE/L2, SINE/Alu, SINE/MIR</td>\n",
       "      <td>0.71</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>10</td>\n",
       "      <td>NTRK1</td>\n",
       "      <td>2</td>\n",
       "      <td>LINE/L2, SINE/Alu</td>\n",
       "      <td>0.47</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>10</td>\n",
       "      <td>PALB2</td>\n",
       "      <td>2</td>\n",
       "      <td>LINE/L2, SINE/Alu</td>\n",
       "      <td>0.47</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>10</td>\n",
       "      <td>PALLD</td>\n",
       "      <td>1</td>\n",
       "      <td>SINE/Alu</td>\n",
       "      <td>0.24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>232 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     labels   gene  counts                     agg_kind  perc  \\\n",
       "0        -1   GLMN       1                     SINE/Alu  0.24   \n",
       "1        -1   GPC3       2            LINE/L1, SINE/Alu  0.47   \n",
       "2        -1    HFE       2            LINE/L1, SINE/Alu  0.47   \n",
       "3        -1   HRAS       2      SINE/Alu, Simple_repeat  0.47   \n",
       "4         0  RTEL1       1                      LINE/L2  0.24   \n",
       "..      ...    ...     ...                          ...   ...   \n",
       "227      10    NF1       2           SINE/Alu, SINE/MIR  0.47   \n",
       "228      10    NF2       3  LINE/L2, SINE/Alu, SINE/MIR  0.71   \n",
       "229      10  NTRK1       2            LINE/L2, SINE/Alu  0.47   \n",
       "230      10  PALB2       2            LINE/L2, SINE/Alu  0.47   \n",
       "231      10  PALLD       1                     SINE/Alu  0.24   \n",
       "\n",
       "                                     similarity_reason  \n",
       "0      Genes related to various syndromes and cancers.  \n",
       "1      Genes related to various syndromes and cancers.  \n",
       "2      Genes related to various syndromes and cancers.  \n",
       "3      Genes related to various syndromes and cancers.  \n",
       "4    Predominantly tumor suppressor genes and genes...  \n",
       "..                                                 ...  \n",
       "227                                                NaN  \n",
       "228                                                NaN  \n",
       "229                                                NaN  \n",
       "230                                                NaN  \n",
       "231                                                NaN  \n",
       "\n",
       "[232 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define similarity reasons for each cluster\n",
    "similarity_reasons = {\n",
    "    -1: \"Genes related to various syndromes and cancers.\",\n",
    "    0: \"Predominantly tumor suppressor genes and genes involved in syndromes or cancer.\",\n",
    "    1: \"Predominantly DNA repair genes and tumor suppressors, associated with various cancers and syndromes.\",\n",
    "    2: \"Genes involved in DNA repair, metabolic disorders, and signal transduction.\",\n",
    "    3: \"Genes involved in DNA repair, apoptosis regulation, and cell cycle control.\",\n",
    "    4: \"Genes involved in DNA repair, tumor suppression, and signaling pathways.\",\n",
    "    5: \"All genes are involved in the Fanconi anemia pathway.\",\n",
    "    6: \"Genes involved in DNA repair, apoptosis regulation, and cell cycle control.\",\n",
    "    7: \"Genes involved in DNA repair, tumor suppression, and signaling pathways.\",\n",
    "    8: \"Genes involved in DNA repair, tumor suppression, and signaling pathways.\"\n",
    "}\n",
    "\n",
    "# Add the new column with similarity reasons to the original dataframe\n",
    "percentages['similarity_reason'] = percentages['labels'].map(similarity_reasons)\n",
    "percentages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae02195-d321-470b-814b-eacdecb91bc2",
   "metadata": {},
   "source": [
    "# Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1484143-46fd-4cc8-87f3-e70a15621cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to create count DataFrame for a given column\n",
    "def create_count_df(column):\n",
    "    count_df = final_df.groupby(by=[column, \"bin_class\"], observed=True).size().to_frame()\n",
    "    count_df.columns = [\"count\"]\n",
    "    count_df = count_df.reset_index()\n",
    "    count_df = count_df.loc[count_df[\"count\"] != 0]\n",
    "    count_df = count_df.sort_values(by=[\"count\"], ascending=False)\n",
    "    return count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d47b89-5259-4f15-859a-0e8ee408d1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['#chrom', 'score',\n",
    "       'blockSizes', 'clinSign',  'type', 'geneId',\n",
    "       'molConseq', 'testedInGtr',\n",
    "       'origin', \"_originCode\", \"_allTypeCode\", 'cytogenetic',\n",
    "       'ClinClass', 'Classification', \"gene\", \"labels\", \"agg_kind\", \"similarity_reason\"]\n",
    "\n",
    "# Initial figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces for each column\n",
    "for col in columns:\n",
    "    count_df = create_count_df(col)\n",
    "    fig.add_trace(go.Bar(\n",
    "        x=count_df[col],\n",
    "        y=count_df['count'],\n",
    "        name=col,\n",
    "        marker_color=count_df['bin_class'],\n",
    "        visible=False  # Initially hide all traces\n",
    "    ))\n",
    "\n",
    "# Show the first trace by default\n",
    "fig.data[0].visible = True\n",
    "\n",
    "# Create the dropdown menu\n",
    "dropdown_buttons = [\n",
    "    {'label': col, 'method': 'update', 'args': [{'visible': [col == trace.name for trace in fig.data]}]}\n",
    "    for col in columns\n",
    "]\n",
    "\n",
    "# Update layout with dropdown\n",
    "fig.update_layout(\n",
    "    updatemenus=[{\n",
    "        'buttons': dropdown_buttons,\n",
    "        'direction': 'down',\n",
    "        'showactive': True\n",
    "\n",
    "    }],\n",
    "    barmode='stack'\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98058e71-2f40-4324-8671-b51543ecfffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "so.Plot(grouped_df, x=\"labels\", color=\"kind\").add(so.Bar(), so.Count(), so.Stack())\n",
    "pivot_df = grouped_df.pivot_table(index='gene', columns='labels', aggfunc='size', fill_value=0)\n",
    "sns.heatmap(pivot_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "429416d4-05c9-4a3d-ae04-50274deff085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create faceted pie charts\n",
    "fig = px.pie(percentages, values='perc', names='gene', facet_col='labels', facet_col_wrap=3, title='Pie Charts of Gene Percentages by Label')\n",
    "fig.update_traces(textinfo=\"none\")\n",
    "\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomics_env",
   "language": "python",
   "name": "genomics_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
