{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8b2a5b6-5cc7-46bd-be48-2fb09166dcc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependency management\n",
    "# --------------------------------------------------------\n",
    "import sys\n",
    "from pathlib import Path\n",
    "project_root = Path(\"..\").resolve()\n",
    "sys.path.append(str(project_root / \"scripts\"))\n",
    "from setup_environment import setup_paths\n",
    "project_root = setup_paths()\n",
    "\n",
    "# Data manipulation\n",
    "# --------------------------------------------------------\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import math\n",
    "import time\n",
    "\n",
    "# Visualizations\n",
    "# --------------------------------------------------------\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import missingno as msno\n",
    "import altair as alt\n",
    "from plotnine import ggplot, geom_point, aes, stat_smooth, facet_wrap\n",
    "from bokeh.plotting import figure, show\n",
    "import seaborn.objects as so\n",
    "import plotly.express as px\n",
    "\n",
    "# Custom scripts\n",
    "# ---------------------------------------------------------\n",
    "from data_processing.cleaner import cleaning, compare_and_drop_duplicates\n",
    "from helper.loader import read_csv\n",
    "\n",
    "# Other\n",
    "# ---------------------------------------------------------\n",
    "from collections import Counter\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bff607bf-b49e-46cd-b940-4a607e958ade",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load raw data\n",
    "var_info = pd.read_csv(f\"{project_root}/data/raw/variation_information.tsv\", sep=\"\\t\")\n",
    "cancer_clinvar = pd.read_csv(f\"{project_root}/data/raw/cancermama_clinvarmain.csv\", sep=\"\\t\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb7004a7-be38-4880-bec6-bf134d12c509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean each dataset\n",
    "var_info = cleaning(var_info, df_name=\"variation_information\")\n",
    "cancer_clinvar = cleaning(cancer_clinvar, df_name=\"cancermama_clinvarmain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4585509c-5e45-450f-8671-029f22d94c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join var_info and cancer_clinvar\n",
    "var_info[\"real_id\"] = var_info[\"rsID\"].astype(str) + \" \" + var_info[\"ClinInfo\"].astype(str)\n",
    "cancer_clinvar[\"real_id\"] = cancer_clinvar[\"snpId\"].astype(str) + \" \" + cancer_clinvar[\"ClinInfo\"].astype(str)\n",
    "cancer_clinvar = cancer_clinvar.drop([\"ClinInfo\"], axis=1)\n",
    "merged_df = cancer_clinvar.merge(var_info, left_on=\"real_id\", right_on=\"real_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f1b9b73-f51c-4c8a-9071-e0ce48293b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns 'clinSign' and 'ClinClass' have 2 differences.\n",
      "Columns '_clinSignCode' and 'Classification' have 2 differences.\n"
     ]
    }
   ],
   "source": [
    "# Remove duplicated columns with the same values\n",
    "merged_df = compare_and_drop_duplicates(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e51b2515-525f-46dc-8f3d-b44af3b6ab2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Binary classification using PG and LP as 1 \n",
    "merged_df['bin_class'] = merged_df['Classification'].apply(lambda x: 1 if x in ['pg'] else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81b7cff2-9948-4ebc-89ef-bc8b0a1c610b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop geneId as it has the same information of gene, _originCode has the same information of origin\n",
    "merged_df.drop([\"geneId\", \"_originCode\", \"_allTypeCode\", \"ClinInfo\"], axis=1, inplace=True)\n",
    "# ids\n",
    "id_columns = [\"numSubmit\", \"_variantId\", \"origName\", \"rcvAcc\", \"snpId\", \"Start\", \"End\"]\n",
    "merged_df.drop(id_columns, axis=1, inplace=True)\n",
    "# url in column, no additional info\n",
    "url_columns = [\"phenotype\"]\n",
    "merged_df.drop(url_columns, axis=1, inplace=True)\n",
    "# _mouseOver have repeated information included in other columns:\n",
    "other_columns = [\"_mouseOver\"]\n",
    "merged_df.drop(other_columns, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bebb8eae-c2da-4563-9e03-265ed9fd4ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill nans per column\n",
    "for i in merged_df.columns:\n",
    "    # 1 missing value in clinSign and ClinClass for the same row\n",
    "    if i in [\"clinSign\", \"ClinClass\", \"molConseq\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"unkwown\")\n",
    "    # 45 missing values in phenotypeList\n",
    "    if i in [\"phenotypeList\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"not provided\")\n",
    "    # 45 missing values in rcvAcc\n",
    "    if i in [\"rcvAcc\"]:\n",
    "        merged_df[i] = merged_df[i].fillna(\"rcv000000000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ac6a09d-775b-457b-a13d-3495aa62ff13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=========================\n",
      "\n",
      "#chrom\n",
      "chr17    49125\n",
      "chr2     30159\n",
      "chr16    29047\n",
      "chr11    24704\n",
      "chr5     24219\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "chromStart\n",
      "37025628    32\n",
      "47414419    32\n",
      "47414420    22\n",
      "95099770    17\n",
      "78131728    16\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "chromEnd\n",
      "37025630    33\n",
      "47414421    28\n",
      "37025608    19\n",
      "95099771    17\n",
      "78131619    16\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "name\n",
      "c>t    49778\n",
      "g>a    47477\n",
      "a>g    40605\n",
      "t>c    30231\n",
      "del    26620\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "score\n",
      "1    222786\n",
      "2     97300\n",
      "3     10212\n",
      "0      8501\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "reserved\n",
      "0,0,128        146341\n",
      "0,210,0        113196\n",
      "210,0,0         49066\n",
      "137,121,212     25602\n",
      "128,128,128      4594\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "blockSizes\n",
      "1    306825\n",
      "2     20478\n",
      "3      2890\n",
      "4      2190\n",
      "5      1184\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "clinSign\n",
      "uncertain significance                          146311\n",
      "likely benign                                    98087\n",
      "pathogenic                                       36150\n",
      "conflicting classifications of pathogenicity     25599\n",
      "benign                                            9485\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "reviewStatus\n",
      "criteria_provided_no_conflict    294495\n",
      "criteria_provided_conflict        25591\n",
      "reviewed                          10212\n",
      "not_classified                     8501\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "type\n",
      "single nucleotide variant    293049\n",
      "deletion                      24727\n",
      "duplication                   10786\n",
      "microsatellite                 3966\n",
      "indel                          3509\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "molConseq\n",
      "missense variant                 153726\n",
      "intron variant                    58942\n",
      "synonymous variant                45246\n",
      "frameshift variant                27494\n",
      "non-coding transcript variant     18329\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "testedInGtr\n",
      "n    338775\n",
      "y        24\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "phenotypeList\n",
      "hereditary cancer-predisposing syndrome    38084\n",
      "not provided                               21836\n",
      "fanconi anemia                             10093\n",
      "not specified                               8202\n",
      "neurofibromatosis,type 1                    7123\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "origin\n",
      "germ       325119\n",
      "unk         10246\n",
      "germsom      1217\n",
      "som           742\n",
      "germnov       405\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "cytogenetic\n",
      "13q13.1     18418\n",
      "11q22.3     16299\n",
      "5q22.2      14170\n",
      "17q21.31    13936\n",
      "17q11.2     13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "vcfDesc\n",
      "chr17:-1:na>na        49\n",
      "chr13:-1:na>na        26\n",
      "chr2:-1:na>na          9\n",
      "chr17:78131665:t>c     8\n",
      "chr17:78132279:g>a     8\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "_clinSignCode\n",
      "vus    146312\n",
      "lb     103710\n",
      "pg      36155\n",
      "cf      25602\n",
      "lp      12911\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "simplified_hgvs\n",
      "c>t, g>a    38780\n",
      "a>g, t>c    23382\n",
      "c>t         19784\n",
      "g>a         19595\n",
      "a>g         18645\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "Gene\n",
      "brca2    18418\n",
      "atm      16299\n",
      "apc      14170\n",
      "brca1    13936\n",
      "nf1      13575\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "ClinClass\n",
      "uncertain significance                          146311\n",
      "likely benign                                    98087\n",
      "pathogenic                                       36150\n",
      "conflicting classifications of pathogenicity     25599\n",
      "benign                                            9485\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "Classification\n",
      "vus    146312\n",
      "lb     103710\n",
      "pg      36155\n",
      "cf      25602\n",
      "lp      12911\n",
      "Name: count, dtype: int64\n",
      "\n",
      "=========================\n",
      "\n",
      "bin_class\n",
      "0    302644\n",
      "1     36155\n",
      "Name: count, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in merged_df.columns:\n",
    "    print(\"=\"*25)\n",
    "    print()\n",
    "    print(merged_df[i].value_counts().nlargest(5))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d11205b0-05bd-495d-9191-ed6a09976c69",
   "metadata": {},
   "source": [
    "# Oncokb classification per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e9065998-5177-4281-9819-d5cbb8089f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#chrom', 'chromStart', 'chromEnd', 'name', 'score', 'reserved',\n",
       "       'blockSizes', 'clinSign', 'reviewStatus', 'type', 'molConseq',\n",
       "       'testedInGtr', 'phenotypeList', 'origin', 'cytogenetic', 'vcfDesc',\n",
       "       '_clinSignCode', 'simplified_hgvs', 'Gene', 'ClinClass',\n",
       "       'Classification', 'bin_class', 'classification_oncokb'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# External data path and files\n",
    "oncokb_df = pd.read_csv(f\"{project_root}/data/external/oncokb_classification.csv\")\n",
    "# String normalization\n",
    "oncokb_df['gen'] = oncokb_df['gen'].str.lower()\n",
    "# Gen classification\n",
    "merged_df = merged_df.merge(oncokb_df, left_on= \"Gene\", right_on=\"gen\", how=\"left\")\n",
    "merged_df = merged_df.drop(['gen','description'], axis=1)\n",
    "merged_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "be2c9737-534e-42d5-8dcb-72b37c8e6800",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "classification_oncokb\n",
      "tumor_supressor     273593\n",
      "oncogen              27616\n",
      "unknown              22058\n",
      "tumor_supressor*      7545\n",
      "oncogen*              4242\n",
      "double_function       3745\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(merged_df['classification_oncokb'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa4e062b-43a0-4831-9a98-49a20a43368a",
   "metadata": {},
   "source": [
    "# Gen label obtained from upgenevsrep clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "17ddc423-3b7a-4581-8190-14eba9457cf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>reserved</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>clinSign</th>\n",
       "      <th>reviewStatus</th>\n",
       "      <th>type</th>\n",
       "      <th>...</th>\n",
       "      <th>bin_class</th>\n",
       "      <th>classification_oncokb</th>\n",
       "      <th>gen</th>\n",
       "      <th>gen_label</th>\n",
       "      <th>string_per_umap_cluster</th>\n",
       "      <th>string_per_umap_cluster_description</th>\n",
       "      <th>string_total_clustering</th>\n",
       "      <th>string_total_clustering_description</th>\n",
       "      <th>x_position</th>\n",
       "      <th>y_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67483073</td>\n",
       "      <td>67483074</td>\n",
       "      <td>c&gt;t</td>\n",
       "      <td>1</td>\n",
       "      <td>0,0,128</td>\n",
       "      <td>1</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>criteria_provided_no_conflict</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>aip</td>\n",
       "      <td>9</td>\n",
       "      <td>-1</td>\n",
       "      <td>not_in_network</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA repair pathways, full network, and HDR thr...</td>\n",
       "      <td>0.215948</td>\n",
       "      <td>0.33245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67483074</td>\n",
       "      <td>67483075</td>\n",
       "      <td>c&gt;g</td>\n",
       "      <td>1</td>\n",
       "      <td>0,210,0</td>\n",
       "      <td>1</td>\n",
       "      <td>benign</td>\n",
       "      <td>criteria_provided_no_conflict</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>aip</td>\n",
       "      <td>9</td>\n",
       "      <td>-1</td>\n",
       "      <td>not_in_network</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA repair pathways, full network, and HDR thr...</td>\n",
       "      <td>0.215948</td>\n",
       "      <td>0.33245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67483135</td>\n",
       "      <td>67483136</td>\n",
       "      <td>a&gt;g</td>\n",
       "      <td>1</td>\n",
       "      <td>0,210,0</td>\n",
       "      <td>1</td>\n",
       "      <td>benign</td>\n",
       "      <td>criteria_provided_no_conflict</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>aip</td>\n",
       "      <td>9</td>\n",
       "      <td>-1</td>\n",
       "      <td>not_in_network</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA repair pathways, full network, and HDR thr...</td>\n",
       "      <td>0.215948</td>\n",
       "      <td>0.33245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67483146</td>\n",
       "      <td>67483147</td>\n",
       "      <td>c&gt;g</td>\n",
       "      <td>1</td>\n",
       "      <td>0,0,128</td>\n",
       "      <td>1</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>criteria_provided_no_conflict</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>aip</td>\n",
       "      <td>9</td>\n",
       "      <td>-1</td>\n",
       "      <td>not_in_network</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA repair pathways, full network, and HDR thr...</td>\n",
       "      <td>0.215948</td>\n",
       "      <td>0.33245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67483153</td>\n",
       "      <td>67483154</td>\n",
       "      <td>g&gt;a</td>\n",
       "      <td>1</td>\n",
       "      <td>0,210,0</td>\n",
       "      <td>1</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>criteria_provided_no_conflict</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>aip</td>\n",
       "      <td>9</td>\n",
       "      <td>-1</td>\n",
       "      <td>not_in_network</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA repair pathways, full network, and HDR thr...</td>\n",
       "      <td>0.215948</td>\n",
       "      <td>0.33245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #chrom  chromStart  chromEnd name  score reserved  blockSizes  \\\n",
       "0  chr11    67483073  67483074  c>t      1  0,0,128           1   \n",
       "1  chr11    67483074  67483075  c>g      1  0,210,0           1   \n",
       "2  chr11    67483135  67483136  a>g      1  0,210,0           1   \n",
       "3  chr11    67483146  67483147  c>g      1  0,0,128           1   \n",
       "4  chr11    67483153  67483154  g>a      1  0,210,0           1   \n",
       "\n",
       "                 clinSign                   reviewStatus  \\\n",
       "0  uncertain significance  criteria_provided_no_conflict   \n",
       "1                  benign  criteria_provided_no_conflict   \n",
       "2                  benign  criteria_provided_no_conflict   \n",
       "3  uncertain significance  criteria_provided_no_conflict   \n",
       "4           likely benign  criteria_provided_no_conflict   \n",
       "\n",
       "                        type  ... bin_class classification_oncokb  gen  \\\n",
       "0  single nucleotide variant  ...         0               unknown  aip   \n",
       "1  single nucleotide variant  ...         0               unknown  aip   \n",
       "2  single nucleotide variant  ...         0               unknown  aip   \n",
       "3  single nucleotide variant  ...         0               unknown  aip   \n",
       "4  single nucleotide variant  ...         0               unknown  aip   \n",
       "\n",
       "  gen_label string_per_umap_cluster string_per_umap_cluster_description  \\\n",
       "0         9                      -1                      not_in_network   \n",
       "1         9                      -1                      not_in_network   \n",
       "2         9                      -1                      not_in_network   \n",
       "3         9                      -1                      not_in_network   \n",
       "4         9                      -1                      not_in_network   \n",
       "\n",
       "  string_total_clustering                string_total_clustering_description  \\\n",
       "0                       1  DNA repair pathways, full network, and HDR thr...   \n",
       "1                       1  DNA repair pathways, full network, and HDR thr...   \n",
       "2                       1  DNA repair pathways, full network, and HDR thr...   \n",
       "3                       1  DNA repair pathways, full network, and HDR thr...   \n",
       "4                       1  DNA repair pathways, full network, and HDR thr...   \n",
       "\n",
       "  x_position y_position  \n",
       "0   0.215948    0.33245  \n",
       "1   0.215948    0.33245  \n",
       "2   0.215948    0.33245  \n",
       "3   0.215948    0.33245  \n",
       "4   0.215948    0.33245  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load labels obtained after clustering upgenevsrep\n",
    "upgenevsrep_clustering_labels = pd.read_csv(f\"{project_root}/data/processed/labels.csv\")\n",
    "upgenevsrep_clustering_labels[\"gen\"] = upgenevsrep_clustering_labels[\"gen\"].str.lower()\n",
    "merged_df = merged_df.merge(upgenevsrep_clustering_labels, left_on=\"Gene\", right_on=\"gen\")\n",
    "merged_df.rename(columns={'label': 'gen_label'}, inplace=True)\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aa04e254-fcda-4fae-9664-534684cd8edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv(f\"{project_root}/data/interim/merged_df.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomics_env",
   "language": "python",
   "name": "genomics_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
